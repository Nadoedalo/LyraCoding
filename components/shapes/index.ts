export { Hexagon } from "./Hexagon";
export { Square } from "./Square";
export { Triangle } from "./Triangle";