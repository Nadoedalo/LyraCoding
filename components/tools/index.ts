import { SelectTool } from "./SelectTool";
import { MoveTool } from "./MoveTool";
import { ClosestPointTool } from "./ClosestPointTool";
import { ExportTool } from "./ExportTool";
export const tools = [SelectTool, MoveTool, ClosestPointTool, ExportTool];