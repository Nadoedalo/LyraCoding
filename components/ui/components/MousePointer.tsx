import {useThree} from "@react-three/fiber";
import * as THREE from "three";
import {useEffect, useRef} from "react";

/**
 * TODO: Work In Progress [WIP]
 * The component that should enable the "closest point tool"
 * I've bashed my head with Raycaster but to no avail.
 * Seems that I miss some basic knowledge about Raycasting
 * But this should be the proper way to do it.
 * Tried to replicate this solution https://jsfiddle.net/f2Lommf5/14316/
 * And failed miserably because I get weird point coordinates
 * Currently everything is commented out to leave traces of failed attempts
 * I always felt that I'm pretty close with the solution
 * */
export const MousePointer = ({mesh, plane, raycaster, index}) => {
    const {size, camera, mouse} = useThree();
    const mouseVector = new THREE.Vector2(0, 0);
    const pointOnPlane = new THREE.Vector3();
    const target = new THREE.Vector3();
    let point = new THREE.Vector3(0, 0, 0);
    const mouseCircle = useRef();
    /*    const position = useMemo(() => {
            if(mousePosition) {
                return [(mouse.x * size.width) / 2.5, (mouse.y * size.height) / 2.5, 1];
            }
            return [0,0,1]
        }, [mousePosition]); */
    useEffect(() => {
        if (!raycaster) {
            return;
        }
        /*const [x, y] = position;
        mouseVector.x = x;
        mouseVector.y = y;*/


        // raycaster.setFromCamera(mouseVector, camera);
        // raycaster.ray.intersectPlane(plane, pointOnPlane);

        //mouseCircle.current.position.copy(pointOnPlane);
        //console.log(pointOnPlane);

        /*const triangle = new THREE.Triangle();
        const geometryPosition = mesh.geometry.attributes.position;

        const { count } = geometryPosition;
        let minDistance = Infinity;
        for(let i = 0; i < count; i += 3) {
            const a = geometryPosition.getX(i);
            const b = geometryPosition.getX(i + 1);
            const c = geometryPosition.getX(i + 2);
            triangle.a.fromBufferAttribute( geometryPosition, a );
            triangle.b.fromBufferAttribute( geometryPosition, b );
            triangle.c.fromBufferAttribute( geometryPosition, c );
            triangle.closestPointToPoint( pointOnPlane, target );

            const distanceSq = pointOnPlane.distanceToSquared( target );
            if ( distanceSq < minDistance ) {

                mouseCircle.current.position.copy(target);
                minDistance = distanceSq;

            }
        }*/
        /* let minDistance = Infinity;
        const minPoint = new THREE.Vector3(0,0,0);
        const vertices = shapeStore.verticesArr[index];
        const length = vertices.length;
        console.log('mesh', mesh)
        let i = 0;
        for(; i < length; i++) {
            const currentVertex = vertices[i];
            const nextVertex = vertices[i + 1] || vertices[0];
            const distance = raycaster.ray.distanceSqToSegment(nextVertex, currentVertex, null, minPoint);
            if(distance < minDistance) {
                point = minPoint;
                minDistance = distance;
                console.log('minIndex', i);
            }
        }*/

    }, [mouse]);
    return /*<mesh ref={mouseCircle}>
            <circleGeometry args={[0.05]} />
            <meshBasicMaterial color="#000000" side={THREE.DoubleSide}/>
        </mesh>*/;
}