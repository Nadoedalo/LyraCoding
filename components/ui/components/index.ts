export { MeshCollection } from "./MeshCollection";
export { MousePointer } from "./MousePointer"