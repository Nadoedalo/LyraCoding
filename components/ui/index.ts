export { ToolBar } from "./ToolBar";
export { Workspace } from "./Workspace";