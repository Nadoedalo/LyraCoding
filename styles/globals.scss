html,
body {
  padding: 0;
  margin: 0;
  font-family: Roboto, Verdana, sans-serif;
  width: 100%;
  height: 100%;

  > div {
    width: 100%;
    height: 100%;
  }
}

main {
  display: flex;
  flex-direction: row;
  width: 100%;
  height: 100%;

  .toolbar {
    background: #03DAC6;
    width: 3em;
    padding: 0.5em;
    display: flex;
    flex-direction: column;
    justify-content: center;
    text-align: center;
    box-sizing: border-box;
    border-right: 2px solid #3700B3;
    > div {
      display: flex;
      flex-direction: column;
      padding-top: 3em;
      font-size: 2em;
      > span {
        padding: 0.25rem 0;
        cursor: pointer;
        display: flex;
        justify-content: center;
        transition: 0.3s ease;
        &:hover, &.active {
          color: #3700B3;
        }
      }
      &:first-child {
        padding-top: 0;
      }
    }
  }

  .workspace {
    background-color: #EFEFEF;
    background-image: radial-gradient(black 1px, transparent 0);
    background-size: 20px 20px;
    flex-grow: 1;
  }
}