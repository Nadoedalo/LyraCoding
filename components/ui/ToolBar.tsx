import { Component } from 'react';
import { ClosestPointTool, MoveTool, SelectTool } from "../tools";
import { Triangle, Square, Hexagon } from "../shapes";

export class ToolBar extends Component<any, any> {
    render() {
        return <div className="toolbar">
            <div className="tools">
                {/*TODO auto-import tools*/}
                <ClosestPointTool />
                <MoveTool />
                <SelectTool />
            </div>
            <div className="shapes">
                <Triangle />
                <Square />
                <Hexagon />
            </div>
        </div>
    }
}