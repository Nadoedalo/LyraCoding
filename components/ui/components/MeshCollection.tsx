import { observer } from "mobx-react-lite";
import { shapeStore } from "../../../store";
import { useThree } from "@react-three/fiber";
import { useLayoutEffect, useRef, useState } from "react";
import * as THREE from "three";
import { useDrag } from "@use-gesture/react";
import { MousePointer } from ".";
import { toolStore } from "../../../store";
import { toolsConfig } from "../../../configs/tools.config";

/**
 * The main collection to draw all the meshes with various shapes
 * Access through shapeIndex enables O(1) access to all the attributes of the shape in the store
 * */
export const MeshCollection = observer(({ store }: { store: shapeStore }) => {
    return (<group>
        {
            store.shapesArr.map((shapeObj, shapeIndex) => (
                <MeshView shape={shapeObj} position={shapeStore.positionArr[shapeIndex]} index={shapeIndex} key={shapeIndex}/>
            ))
        }
    </group>);
});

/**
 * The main Mesh View. Could be in different module but MeshCollection doesn't have a lot of logic yet
 * Does:
 * Selecting, Dragging
 * And supposedly should provide support elements for Closest Point Tool
 * FIXME separate module? Seems unnecessary due to low logic in MeshCollection
 * */
const MeshView = observer(({ shape, index, position }) => {
    const {size, viewport} = useThree();
    const aspect = size.width / viewport.width;
    const [hover, setHover] = useState(false);
    const geometry = useRef();
    const mesh = useRef();
    const outline = useRef();
    const plane = useRef();
    const raycaster = useRef();
    const dot = new THREE.Vector3(0, 0, 1);
    const dragBind = useDrag((state) => {
        state.event.stopPropagation();
        if(toolStore.activeTools[toolsConfig.MOVE_TOOL]) {
            const {offset: [x, y]} = state;
            const position = [x / aspect, -y / aspect, 0]
            shapeStore.updatePosition(index, position);
        }
    }, {});
    useLayoutEffect(() => {
        if (outline.current && geometry.current) {
            outline.current.geometry = new THREE.EdgesGeometry(geometry.current);
        }
    }, [geometry])
    return (<group position={position}>
        {/*Actual shape*/}
        <mesh
            key={index}
            {...dragBind()}
            onClick={(e) => {
                e.stopPropagation();
                if(toolStore.activeTools[toolsConfig.SELECT_TOOL]) {
                    shapeStore.selectShape(shape);
                }
            }}
            onPointerOver={(e) => { // FIXME: Glitchy because it isn't tied through Raycaster but it should
                e.stopPropagation();
                if (!hover) {
                    setHover(true)
                }
            }}
            onPointerOut={(e) => {
                e.stopPropagation();
                if (hover) {
                    setHover(false)
                }
            }}
            ref={mesh}

        >
            <shapeGeometry args={[shape.shape]} ref={geometry}/>
            <meshBasicMaterial
                color={shape.active || hover ? '#03DAC6' : '#A3D951'}
            />
        </mesh>
        {/*Outline*/}
        <mesh>
            <lineSegments ref={outline}>
                <meshBasicMaterial color="#3700B3" visible={shape.active}/>
            </lineSegments>
        </mesh>
        {/*Raycaster, Plane & MousePointer necessary for the Closest Point Tool*/}
        <raycaster
            ref={raycaster}
        ></raycaster>
        <plane args={[dot, 0]} ref={plane}/>
        <MousePointer
            mesh={mesh.current}
            plane={plane.current}
            raycaster={raycaster.current}
            index={index}
        />
    </group>)
})