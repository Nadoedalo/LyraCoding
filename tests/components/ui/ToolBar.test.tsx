import renderer from 'react-test-renderer';
import { describe, expect, it } from "@jest/globals";
import { ToolBar } from "../../../components/ui";

describe('ToolBar Snapshot', () => {
    const toolBar = renderer.create(<ToolBar />,);
    let tree = toolBar.toJSON();
    it("Should match Snapshot", () => {
        expect(tree).toMatchSnapshot();
    })
})