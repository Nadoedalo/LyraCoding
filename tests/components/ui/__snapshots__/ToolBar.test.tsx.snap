// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolBar Snapshot Should match Snapshot 1`] = `
<div
  className="toolbar"
>
  <div
    className="tools"
  >
    <span
      className="active"
      onClick={[Function]}
      title="Select Tool"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M123.193 29.635L121 406.18l84.31-82.836 65.87 159.02 67.5-27.96-65.87-159.02L391 294.342z"
        />
      </svg>
    </span>
    <span
      className="active"
      onClick={[Function]}
      title="Move Tool"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 11h-5V6h3l-4-4-4 4h3v5H6V8l-4 4 4 4v-3h5v5H8l4 4 4-4h-3v-5h5v3l4-4-4-4z"
        />
      </svg>
    </span>
    <span
      className=""
      onClick={[Function]}
      title="Closest Point"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17 11H7V7l-5 5 5 5v-4h10v4l5-5-5-5z"
        />
      </svg>
    </span>
    <span
      className=""
      onClick={[Function]}
      title="Export to console"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 16h2V7h3l-4-5-4 5h3z"
        />
        <path
          d="M5 22h14c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2h-4v2h4v9H5v-9h4V9H5c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2z"
        />
      </svg>
    </span>
  </div>
  <div
    className="shapes"
  >
    <span
      onClick={[Function]}
      title="Add Triangle"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.022 1.566a1.13 1.13 0 0 1 1.96 0l6.857 11.667c.457.778-.092 1.767-.98 1.767H1.144c-.889 0-1.437-.99-.98-1.767L7.022 1.566z"
          fillRule="evenodd"
        />
      </svg>
    </span>
    <span
      onClick={[Function]}
      title="Add Square"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M512 512H0V0h512v512z"
        />
      </svg>
    </span>
    <span
      onClick={[Function]}
      title="Add Hexagon"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M17.2 3H6.8l-5.2 9 5.2 9h10.4l5.2-9z"
        />
      </svg>
    </span>
  </div>
</div>
`;
